### DISCUSSION TITLE
Use of Pagination Token for Cursory Support of Pagination Results

### DISCUSSION THREAD

**Discussion by patrickjennings** [keycloak/keycloak#28250] [Ideas]
The use of pagination tokens / cursors, instead of offset / first, as a means of specifying and retrieving the next pagination API result set, is something that many large scale APIs are trending towards. See:
* https://slack.engineering/evolving-api-pagination-at-slack/
* https://developers.google.com/youtube/v3/guides/implementation/pagination 

Keycloak currently uses "first" and "max" in a number of endpoints, in order to specify the first position of the result and number of results to return. The issue is that this:
1.  Does not scale well with very large data sets. Each subsequent query requires the DB to scan through indices / tables especially with large offsets provided.
2. The results returned may not be consistent. When instances are removed / added to the table, the results may be duplicated when the next page is queried for.

A pagination cursor works by providing a token back to the client on the first query. The token is used to get the next page of results. In the backend, the token translates to a unique identifier, for example the primary key of the table, of where to start the next page.

This allows efficient use of indices on the table:
```sql
SELECT id
  FROM client
 ORDER BY id
OFFSET 5000 ROWS
 FETCH FIRST 50 rows only;

-- Becomes

 SELECT id
  FROM client
 WHERE ID > 'acce35e8-7e99-4d6e-bc17-69485fd6493a'
 ORDER BY id
 FETCH FIRST 50 rows only;
```

Benefits include:
* Help stabilize the performance when having to iterate over and offset into a large amount of data.
* Keeps results consistent in that the same result will not be duplicated in further calls to paginated results.

There are a few things to note:
* "Seeking" behavior would not be supported. Querying to see, for example, whether there is an entity present at `x` position would not be possible.
* Ideally the ID tracked would be sequential. Our primary keys are type 4 UUIDs which are not sequential (could look into UUID7 for sequential UUIDs in the future). This means that there would be a chance that newly created entities may be skipped when iterating over the pages.

This initiative could be covered in:
* https://github.com/keycloak/keycloak/discussions/8983

We can utilize the Link header to automatically point to the next page results. Client should be instructed to use the next Link header in the HTTP response to query the next page:
* https://github.com/keycloak/keycloak-community/blob/main/design/rest-api-guideline.md#pagination

Here is a performance breakdown of queries iterating through clients with a particular client attribute. You can see the impact the offset has on the query performance:
* https://gist.github.com/patrickjennings/64d1b71e2c16f5fddf18c2bef00bc529

This individuals problem may well be this issue as well:
* https://keycloak.discourse.group/t/paged-reading-of-users-via-get-realm-users-gets-progressively-slower/7447

**Comment by sschu** [comment]
I think this is a good idea. Would you also have time to work on this?

**Comment by danielFesenmeyer** [comment]
@patrickjennings I have a question: Why should the ID tracked be sequential? I think there are several cases where you like to sort by a name instead of an ID (for example when querying users). How does a sequential ID help in this case?

**Reply by danielFesenmeyer** [reply]
From my understanding, it's just important that the order criteria is backed by some kind of index.
